posthog/api/query.py

```diff
@@ -196,56 +196,78 @@ def _tag_client_query_id(self, query_id: str | None):
         set_tag("client_query_id", query_id)
 
 
-ASYNC_FALLBACK_TO_POLLING_TIMEOUT = 20
+MAX_QUERY_TIMEOUT = 600
 
 
-async def query_awaited(request: Request, *args, **kwargs) -> HttpResponse:
-    """Async endpoint for handling event source queries."""
-
+async def query_awaited(request: Request, *args, **kwargs) -> StreamingHttpResponse:
+    """Async endpoint for handling event source queries using Server-Sent Events (SSE)."""
     # Call the create method on QueryViewSet
     view = await sync_to_async(QueryViewSet.as_view)({"post": "create"}, **kwargs)
     response = await sync_to_async(view)(request)
 
-    if (
-        response.status_code != 202
-    ):  # 202 means accepted, which means we're calculating async. Anything else means we can just return immediately
-        return response
+    if response.status_code != 202:  # Non-202 means we can return immediately
+        response.render()
+        content = response.rendered_content.decode("utf-8")
+        return StreamingHttpResponse(
+            [f"data: {content}\n\n".encode()],
+            status=response.status_code,
+            content_type="text/event-stream",
+            headers={
+                "Cache-Control": "no-cache",
+                "X-Accel-Buffering": "no",
+            },
+        )
 
     response.render()
     data = json.loads(response.rendered_content)
 
-    # For async responses, poll until complete or timeout
-    async def check_query_status():
+    async def event_stream():
         assert kwargs.get("team_id") is not None
         manager = QueryStatusManager(data["query_status"]["id"], cast(int, kwargs["team_id"]))
         start_time = time.time()
-        sleep_time = 0.1  # Start with 100ms
-        max_sleep_time = 1.0  # Don't wait more than 1 second between checks
+        last_update_time: float = start_time
+
+        # For things to feel snappy we want to frequently check initially, then back off so we don't overload redis
+        FAST_POLL_DURATION = 3.0  # First 3 seconds
+        MEDIUM_POLL_DURATION = 15.0  # Until 15 seconds
+        FAST_POLL_INTERVAL = 0.05
+        MEDIUM_POLL_INTERVAL = 0.1
+        SLOW_POLL_INTERVAL = 1.0
```

Comment:
Feels like typical exponential backoff would be simpler than this tiering, and more aligned with how humans perceive time, i.e. logarithmically (e.g. polling every 50ms matters when we've been waiting just 150ms, but after 2500ms, 50ms is not perceptibly different from 100ms)